blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md

```diff
@@ -0,0 +1,238 @@
+  of the first goal).
+
+- If a proof step now proves *more* things, it is also very bad! The
+  next proof steps, meant for the first subgoal (for example), would
+  then apply to the beginning of the second subgoal, and you get very
+  confusing errors again.
+
+What we need for robustness is a way to indicate our *intent* to Coq,
+when we think that a subgoal is finished and that a new subgoal
+starts, so that Coq can fail loudly at the moment where it notices
+that this intent does not match reality, instead of at an arbitrary
+later time.
+
+(The `S*Case` tactics used in Software Foundations can solve this
+problem if used in a carefuly, systematic way, and additionally
+provides naming. Alexandre Pilkiewicz implemented an even more
+powerful [cases](https://github.com/pilki/cases) plugin. Bullets are
+available in standard Coq since 8.4 (released in 2012), and can be
+used with no effort.)
+
+There is not much discussion of bullets around; see the
+[documentation](https://coq.inria.fr/distrib/8.6/refman/Reference-Manual009.html#sec326)
+in the Coq manual. I learned a lot from Arthur Azevedo de
+Amorim's
+[Bullets.v](https://github.com/arthuraa/poleiro/blob/master/theories/Bullets.v)
+file.
+
+Finally, some people don't use bullets, because they systematically
+use so much automation that they never see subgoals -- each lemma has
+a one-line proof. This is also a valid style. (I have been going to
+Adam Chlipala's [Formal Reasoning about
+Programs](https://frap.csail.mit.edu/main) 2017 class, where Adam
+ignores bullets because that is his usual style.) Because I am not
+crushy enough to do this from the start, my proofs tend to start with
+cases and subgoals, and then I refine them to add more automation for
+robustness. I found bullets very useful for the first step, and during
+the refinement process.
+
+# Bullets
+
+Bullets are actually a combination of two features, braces `{ ... }`
+and actually list bullets, `-`, `+`, or `*` or homogeneous repetitions
+of those, for example `--` or `***`.
+
+## Braces
+
+The opening brace `{` focuses the proof on the current subgoal. If you
+finish the proof of the subgoal, the following subgoal will not become
+accessible automatically; you have to use the closing brace `}`
+first. (If you finish the goal earlier than you think, you get an
+error.) Conversely, `}` fails if the subgoal is not complete. (If you
+fail to finish, you get an error.)
+
+The previous example can thus be written as follows, and will be more robust:
+
+```
+induction foo. (* this creates many subgoal *)
+{
+  proof of first subgoal.
+}
+{
+  proof of second subgoal.
+}
+```
+
+If you also want to make sure that the number of subgoals does not
+change, you can use an outer layer of braces:
+
+
+```
+{ induction foo. (* this creates many subgoal *)
+  {
+    proof of first subgoal.
+  }
+  {
+    proof of second subgoal.
+  }
+} (* would fail if a new subgoal appeared *)
```

Comment:
The idea is that you may add a constructor to the inductive, which would create a new subgoal in each induction. I will try to clarify that before.